File: web/console/src/lib/record-path.test.ts
/** The console's half of `internal/vocabulary`'s record-path grammar. The cases
 * are the Go suite's (`TestSplitRecordPath`), deliberately: the two splits are
 * a mirror, and a case that passes there and fails here is the drift this file
 * exists to catch. */

import { describe, expect, it } from "vitest"

import { coerceReferencePath, recordPath, splitRecordPath } from "./record-path"

describe("splitRecordPath", () => {
  it("splits a qualified kind's path on the kind grammar, not a registry", () => {
    expect(
      splitRecordPath("substrate.reamde.dev/core/llmprovider/claude")
    ).toEqual({ kind: "substrate.reamde.dev/core/llmprovider", id: "claude" })
  })

  it("refuses a dotless first segment: every kind carries an authority", () => {
    expect(splitRecordPath("task/abc123")).toBeUndefined()
  })

  it("keeps the slashes inside an id: a declaration id IS a kind reference", () => {
    expect(
      splitRecordPath(
        "substrate.reamde.dev/core/kind/samples.substrate.reamde.dev/tasks/task"
      )
    ).toEqual({
      kind: "substrate.reamde.dev/core/kind",
      id: "samples.substrate.reamde.dev/tasks/task",
    })
    expect(splitRecordPath("substrate.reamde.dev/core/note/a/b/c")).toEqual({
      kind: "substrate.reamde.dev/core/note",
      id: "a/b/c",
    })
  })

  it("refuses everything that is not a path, so a pin can complete it", () => {
    // The authored short form: a bare id, and the bare id of a DECLARATION,
    // which looks like a path but has nothing left after its kind.
    expect(splitRecordPath("claude")).toBeUndefined()
    expect(
      splitRecordPath("samples.substrate.reamde.dev/tasks/task")
    ).toBeUndefined()
    expect(splitRecordPath("")).toBeUndefined()
    expect(splitRecordPath("/claude")).toBeUndefined()
    expect(splitRecordPath("task/")).toBeUndefined()
    expect(splitRecordPath("substrate.reamde.dev/core/")).toBeUndefined()
  })

  it("round-trips through recordPath", () => {
    for (const path of [
      "substrate.reamde.dev/core/llmprovider/claude",
      "substrate.reamde.dev/core/note/a/b/c",
      "substrate.reamde.dev/core/kind/samples.substrate.reamde.dev/tasks/task",
    ]) {
      const parts = splitRecordPath(path)
      expect(parts && recordPath(parts.kind, parts.id)).toBe(path)
    }
  })
})

/** The console's half of the write path's own coercion
 * (`engine.coerceReferencePath`). The cases are the ones the hardening added:
 * a value that reads two ways is refused naming BOTH, an empty-segment shape
 * names no record, and a slash-bearing value that does not parse as a path is
 * still the ordinary short form, which is what keeps an identity spellable. */
describe("coerceReferencePath", () => {
  const FUNCTION = "substrate.reamde.dev/core/function"
  const KIND = "substrate.reamde.dev/core/kind"

  it("leaves a full path under its own pin alone", () => {
    expect(
      coerceReferencePath(
        FUNCTION,
        `${FUNCTION}/substrate.reamde.dev/core/graphql`
      )
    ).toEqual({ value: `${FUNCTION}/substrate.reamde.dev/core/graphql` })
  })

  it("completes a bare id from the pin", () => {
    expect(
      coerceReferencePath("substrate.reamde.dev/core/llmprovider", "claude")
    ).toEqual({
      value: "substrate.reamde.dev/core/llmprovider/claude",
    })
  })

  it("keeps a slash-bearing SHORT FORM, because it parses as no path", () => {
    // `substrate.reamde.dev/core/graphql` has a dotted first segment and
    // nothing left after its kind, so it cannot be read as a path: it is the
    // identity, and the pin completes it. This is what a tool entry writes.
    expect(
      coerceReferencePath(FUNCTION, "substrate.reamde.dev/core/graphql")
    ).toEqual({ value: `${FUNCTION}/substrate.reamde.dev/core/graphql` })
    expect(coerceReferencePath(KIND, "web.example.com/page")).toEqual({
      value: `${KIND}/web.example.com/page`,
    })
  })

  it("REFUSES a value that reads two ways, naming both readings", () => {
    const problem = coerceReferencePath(
      "p.example.com/p/target",
      "foo.bar/baz/qux/quux"
    )
    expect(problem.value).toBeUndefined()
    expect(problem.error).toContain("ambiguous")
    // Both ends, so either can be the one to change.
    expect(problem.error).toContain("foo.bar/baz/qux")
    expect(problem.error).toContain(
      "p.example.com/p/target/foo.bar/baz/qux/quux"
    )
  })

  it("refuses an empty-segment shape, which names no record", () => {
    // Neither parses as a path, so each is read as a short form, and a short
    // form with a slash on an end is an id of nothing.
    for (const bad of ["target/", "/x"]) {
      expect(coerceReferencePath(FUNCTION, bad).value).toBeUndefined()
      expect(coerceReferencePath(FUNCTION, bad).error).toMatch(
        /has an empty segment/
      )
    }
    // A full path under its own pin whose id half is empty.
    expect(coerceReferencePath(KIND, `${KIND}//b`).error).toMatch(
      /empty id segment/
    )
    // Under any OTHER pin it is the ambiguous case, exactly as the engine
    // reads it: a pointer at the parsed kind, or an id the pin would complete.
    expect(coerceReferencePath(FUNCTION, `${KIND}//b`).error).toMatch(
      /ambiguous/
    )
  })

  it("unpinned, takes a full path and nothing else", () => {
    // Every kind carries an authority and a package, so the qualified path is
    // the only unpinned value that names a record.
    expect(coerceReferencePath("", "foo.bar/baz/qux/abc")).toEqual({
      value: "foo.bar/baz/qux/abc",
    })
    // A kind identity leaves nothing for an id; a dotless first segment and a
    // bare id are no path at all.
    expect(coerceReferencePath("", "foo.bar/baz/qux").error).toMatch(
      /is not one/
    )
    expect(coerceReferencePath("", "note/abc").error).toMatch(/is not one/)
    expect(coerceReferencePath("", "abc").error).toMatch(/is not one/)
    expect(coerceReferencePath("", "foo.bar/baz/qux//abc").error).toMatch(
      /empty id/
    )
  })

  it("says an empty value needs an id", () => {
    expect(coerceReferencePath(FUNCTION, "").error).toBe(
      "a reference needs an id"
    )
  })
})

File: internal/vocabulary/recordpath_test.go
package vocabulary_test

import (
	"testing"

	"github.com/geoah/substrate/internal/vocabulary"
)

// THE GRAMMAR INVARIANT SplitRecordPath rests on: an authority always carries a
// dot, a package and a kind NAME never do. Everything else in this file assumes
// it, and a validator that ever admitted a dotless authority or a dotted name
// would make a flat reference value ambiguous — silently, because the parse
// would keep answering, just wrongly. So it is asserted against the validators
// themselves rather than restated in a comment.
func TestKindGrammarSeparatesAuthorityFromName(t *testing.T) {
	for _, s := range []string{"core", "substrate", "a", "z9", "core-x"} {
		if vocabulary.ValidAuthority(s) {
			t.Errorf("ValidAuthority(%q) = true: a dotless authority makes <kind>/<id> ambiguous", s)
		}
	}
	for _, s := range []string{"substrate.reamde.dev", "example.com", "a.b"} {
		if !vocabulary.ValidAuthority(s) {
			t.Errorf("ValidAuthority(%q) = false", s)
		}
	}
	for _, s := range []string{"core", "tasks", "llm2"} {
		if !vocabulary.ValidPackage(s) {
			t.Errorf("ValidPackage(%q) = false", s)
		}
	}
	for _, s := range []string{"my.package", "Core", "core-x", "core/x"} {
		if vocabulary.ValidPackage(s) {
			t.Errorf("ValidPackage(%q) = true: a package is one lowercase word", s)
		}
	}
	for _, s := range []string{"task", "llmprovider", "person2"} {
		if !vocabulary.ValidName(s) {
			t.Errorf("ValidName(%q) = false", s)
		}
	}
	for _, s := range []string{"my.kind", "a.b", "task/sub", "Task"} {
		if vocabulary.ValidName(s) {
			t.Errorf("ValidName(%q) = true: a dotted or slashed kind name makes <kind>/<id> ambiguous", s)
		}
	}
}

func TestSplitRecordPath(t *testing.T) {
	for name, tc := range map[string]struct {
		path string
		kind string
		id   string
		ok   bool
	}{
		"qualified kind": {
			path: "substrate.reamde.dev/core/llmprovider/claude",
			kind: "substrate.reamde.dev/core/llmprovider", id: "claude", ok: true,
		},
		// A declaration record's id is a kind reference, so its path has four
		// segments and the id keeps its own slash.
		"declaration record id": {
			path: "substrate.reamde.dev/core/kind/samples.substrate.reamde.dev/tasks/task",
			kind: "substrate.reamde.dev/core/kind", id: "samples.substrate.reamde.dev/tasks/task", ok: true,
		},
		"qualified kind, id with slashes": {
			path: "substrate.reamde.dev/core/note/a/b/c",
			kind: "substrate.reamde.dev/core/note", id: "a/b/c", ok: true,
		},
		// Every kind carries an authority (decision 0042), so a dotless first
		// segment is no stored path: a bare "<name>/<id>" like "task/abc123"
		// never names a stored record.
		"dotless first segment is not a path": {path: "task/abc123"},
		// The authored short form of a reference to a declaration: a bare id
		// that LOOKS like a path but has no id left over. It must not parse, or
		// the pin could never complete it.
		"bare declaration id is not a path": {path: "samples.substrate.reamde.dev/tasks/task"},
		"bare id":                           {path: "claude"},
		"empty":                             {path: ""},
		"leading slash":                     {path: "/claude"},
		"trailing slash":                    {path: "task/"},
		"authority alone":                   {path: "substrate.reamde.dev/core/"},
	} {
		t.Run(name, func(t *testing.T) {
			kind, id, ok := vocabulary.SplitRecordPath(tc.path)
			if ok != tc.ok || kind != tc.kind || id != tc.id {
				t.Fatalf("SplitRecordPath(%q) = (%q, %q, %v), want (%q, %q, %v)",
					tc.path, kind, id, ok, tc.kind, tc.id, tc.ok)
			}
			if tc.ok && vocabulary.RecordPath(kind, id) != tc.path {
				t.Errorf("RecordPath does not round-trip %q", tc.path)
			}
		})
	}
}
